### PR TITLE
Callback triggering a scan after receiving a `startAgentScan` message.

### PR DESCRIPTION
**Pull request by deadly-panda**

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -425,7 +429,7 @@ def vulnerability_location_file(
 
 
 @pytest.fixture()
-def data_start_agent_scan() -> dict[str:Any]:
+def data_start_agent_scan() -> Dict[str, Any]:
```
Comment (3asm) [inline tests/conftest.py:432]:
```suggestion
def data_start_agent_scan() -> dict[str, Any]:
```

File: src/ostorlab/scanner/start.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 """Module Responsible for subscribing to nats for different subjects."""
 import logging
 import asyncio
+import functools
```
Comment (3asm) [inline src/ostorlab/scanner/start.py:4]: I don't find the filename to be helpful.

File: src/ostorlab/scanner/callbacks.py [hunk 3]
```diff
@@ -1,6 +1,160 @@
...
+            mask=ip_request.mask or str(ip_network.prefixlen),
+        )
+    else:
+        raise ValueError(f"Invalid Ip address {ip_request.host}")
+
+
+def _asset_instances_from_request(request) -> List[asset.Asset]:
+    assets = []
+    asset_type = request.WhichOneof("asset")
+    if asset_type in ("ip", "ip4v", "ipv6"):
```
Comment (3asm) [inline src/ostorlab/scanner/callbacks.py:56]: Can you please use pattern matching for improved readability.

File: src/ostorlab/scanner/callbacks.py [hunk 4]
```diff
@@ -1,6 +1,160 @@
...
+        logger.error(
+            "%s not supported from scan reference  %s ",
+            asset_type,
+            request.reference_scan_id,
+        )
+
+    return assets
+
+
+def start_scan(subject, request, state_reporter):
```
Comment (3asm) [inline src/ostorlab/scanner/callbacks.py:138]: Why pass subject if it is not used?

File: src/ostorlab/scanner/callbacks.py [hunk 5]
```diff
@@ -1,6 +1,160 @@
...
+        logger.error(
+            "The runtime does not support the provided agent list or group definition."
+        )
+    _install_agents(runtime_instance, request.agents)
+
+    runtime_instance.scan(
+        agent_group_definition=agent_group_definition,
+        assets=assets,
+        title=None,
+    )
```
Comment (3asm) [inline src/ostorlab/scanner/callbacks.py:160]: This code needs a but more restructuring. It should read something like:

```python
agent_group_definition = _extract_agent_group_definition(request)
assets = _extract_assets(request)
scan_id = _extract_scan_id(request)
_update_state_reporter(request)

runtime = _select_runtime(request)
if _can_run(runtime, agent_group_definition) is False:
   ...
else:
  _install_agents(...)
  _scan(...)

```

File: tests/runtimes/local/runtime_test.py [hunk 6]
```diff
@@ -182,3 +186,43 @@ def docker_services():
...
+    assert start_agent_mock_call_args[0][0][0].key == "agent/ostorlab/agent42"
+    assert start_agent_mock_call_args[1][0][0].key == "agent/ostorlab/inject_asset"
+    assert all(
+        call_arg[0][0].key != "agent/ostorlab/local_persist_vulnz"
+        for call_arg in start_agent_mock_call_args
+    )
+    assert all(
+        call_arg[0][0].key != "agent/ostorlab/tracker"
+        for call_arg in start_agent_mock_call_args
+    )
```
Comment (3asm) [inline tests/runtimes/local/runtime_test.py:228]: Explicit checks.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: src/ostorlab/scanner/callbacks.py (hunk 3 above)
Comment (deadly-panda) [reply src/ostorlab/scanner/callbacks.py:56]: Are we dropping support for py3.9 any time soon?

**Comment by 3asm** [comment]
Ping.

**Comment by deadly-panda** [comment]
> Ping.

Please, kindly re-review!

**Review by amine3 — DISMISSED** [review]

**Review by 3asm — APPROVED** [review]

**Pull request merged**